### DISCUSSION TITLE
Data and pointer section, and some terms about encoding

### DISCUSSION THREAD

**Discussion by yuyuankang** [capnproto/capnproto#1871] [Q&A]
Hi, 

I'm new to Cap'n Proto and curious about its binary encoding. While the [encoding documentation](https://capnproto.org/encoding.html) is helpful, I'm still unclear about the definition of "data section" and "pointer section".

I've written some Cap'n Proto and C++ code to explore this:

``` capnp
struct Person {
  id @0 :UInt32;
  name @1 :Text;
  email @2 :Text;
  phones @3 :List(PhoneNumber);

  struct PhoneNumber {
    number @0 :Text;
    type @1 :Type;
  }

  union {
    unemployed @4 :Void;
    employer @5 :Text;
    school @6 :Text;
    selfEmployed @7 :Void;
  }
}

enum Type {
    mobile @0;
    home @1;
    work @2;
}
```

``` c++
    int fd = open("addressbook.bin", O_CREAT | O_WRONLY);
    capnp::MallocMessageBuilder message;
    Person::Builder person = message.initRoot<Person>();
    person.setId(123);
    person.setName("Alice");
    person.setEmail("alice@gmail.com");
    auto alicePhones = person.initPhones(1);
    alicePhones[0].setNumber("555-1212");
    alicePhones[0].setType(Type::HOME);
    person.setSchool("School");
    capnp::writeMessageToFd(fd, message);
    close(fd);
```

In the resulting addressbook.bin file, I assume the first word to be a struct pointer, as it is said in the document that "The first word of the first segment of the message is always a pointer pointing to the message’s root struct."

```
lsb                      struct pointer                       msb
+-+-----------------------------+---------------+---------------+
|A|             B               |       C       |       D       |
+-+-----------------------------+---------------+---------------+

A (2 bits) = 0, to indicate that this is a struct pointer.
B (30 bits) = Offset, in words, from the end of the pointer to the
    start of the struct's data section.  Signed.
C (16 bits) = Size of the struct's data section, in words.
D (16 bits) = Size of the struct's pointer section, in words.
```

However, the first word in the file is 

```
00 00 00 00 0F 00 00 00
```

This implies C = 15 and D = 0, suggesting "the size of pointer section" is 0, yet I expected some pointer for the phones list. Am I missing something?

Additionally, I can't find something as a "list pointer" in the binary file.

**Comment by kentonv** [comment]
When you serialize the message, it is prefixed with a segment table, which comes before the first segment. The 8 bytes you are looking at here are the segment table -- they indicate that there is one segment of 15 words. The _next_ 8 bytes in this case would the the first word of the first segment, where the root pointer lives.

**Reply by lucagebhardt** [reply]
Hello @kentonv, I have a question about the same line in the documentation. I believe it's just a matter of wording, but I want to make sure I understand it correctly.

Since an arbitrary number of objects can exist in a segment, it is possible to have just for example a List as the only object in my segment. In that case, would the serializer wrap the list in a struct and add let the root pointer to that struct, which then points to the list?

In other words, would the layout look like this: `root pointer -> struct pointer -> list pointer`, where the struct pointer is the extra object added by the serializer?

**Reply by kentonv** [reply]
Only the first segment has a root pointer.

By convention, the root pointer of a message is supposed to be a struct, not a list. However, technically you can serialize messages with any kind of object as the root.

**Reply by lucagebhardt** [reply]
Thank you for your answer. I’m still struggling to understand why the first segment requires a pointer to the root struct.

**Reply by kentonv** [reply]
Cap'n Proto pointers encode type information about the thing they point to. In the case of structs, this includes the size of the struct. So without a pointer, it would be impossible to interpret the struct content, because the system wouldn't know where it ends.
